internal/common/function/function.go

```diff
@@ -157,6 +167,82 @@ func ValidateRestoreSpec(ctx context.Context, clientInstance client.Client, nonA
 	return nil
 }
 
+// ValidateBslSpec return nil, if NonAdminRestore is valid; error otherwise
+func ValidateBslSpec(ctx context.Context, clientInstance client.Client, nonAdminBsl *nacv1alpha1.NonAdminBackupStorageLocation, enforcedBslSpec *velerov1.BackupStorageLocationSpec) error {
+	if nonAdminBsl.Spec.BackupStorageLocationSpec.Credential == nil {
+		return fmt.Errorf("NonAdminBackupStorageLocation spec.bslSpec.credential is not set")
+	} else if nonAdminBsl.Spec.BackupStorageLocationSpec.Credential.Name == constant.EmptyString || nonAdminBsl.Spec.BackupStorageLocationSpec.Credential.Key == constant.EmptyString {
+		return fmt.Errorf("NonAdminBackupStorageLocation spec.bslSpec.credential.name or spec.bslSpec.credential.key is not set")
+	}
+
+	var enforcedSpec reflect.Value
+	if enforcedBslSpec != nil {
+		// Ensure that enforcedBslSpec is a pointer
+		enforcedSpec = reflect.ValueOf(enforcedBslSpec)
+		if enforcedSpec.Kind() == reflect.Ptr {
+			enforcedSpec = enforcedSpec.Elem()
+		}
+	}
+
+	// Check if enforcedSpec is valid
+	if enforcedSpec.IsValid() {
+		// Ensure that nonAdminBsl.Spec.BackupStorageLocationSpec is a pointer
+		currentSpec := reflect.ValueOf(nonAdminBsl.Spec.BackupStorageLocationSpec)
+		if currentSpec.Kind() == reflect.Ptr {
+			currentSpec = currentSpec.Elem()
+		}
+		for index := 0; index < enforcedSpec.NumField(); index++ {
+			enforcedField := enforcedSpec.Field(index)
+			enforcedFieldName := enforcedSpec.Type().Field(index).Name
+			currentField := currentSpec.FieldByName(enforcedFieldName)
+
+			// Special handling for Config map
+			// User should be able to set any Config field, but not override the enforced field
+			if enforcedFieldName == "Config" {
+				enforcedConfig, ok := enforcedField.Interface().(map[string]string)
+				if !ok {
+					return fmt.Errorf("failed to cast enforcedField to map[string]string")
+				}
+				currentConfig, ok := currentField.Interface().(map[string]string)
+				if !ok {
+					return fmt.Errorf("failed to cast currentField to map[string]string")
+				}
+				for key, enforcedValue := range enforcedConfig {
+					if currentValue, exists := currentConfig[key]; exists && !reflect.DeepEqual(enforcedValue, currentValue) {
+						return fmt.Errorf(
+							"NonAdminBackupStorageLocation spec.bslSpec.config.%v field value is enforced by admin user, cannot override it",
+							key,
+						)
+					}
+				}
+				continue
+			}
+
+			if !enforcedField.IsZero() && !currentField.IsZero() && !reflect.DeepEqual(enforcedField.Interface(), currentField.Interface()) {
+				field, _ := currentSpec.Type().FieldByName(enforcedFieldName)
+				tagName, _, _ := strings.Cut(field.Tag.Get(constant.JSONTagString), constant.CommaString)
+				return fmt.Errorf(
+					"NonAdminBackupStorageLocation spec.bslSpec.%v field value is enforced by admin user, cannot override it",
+					tagName,
+				)
+			}
+		}
+	}
+
+	// Check if the secret exists in the same namespace
+	secret := &corev1.Secret{}
```

Comment:
Should we also check for the `key` specified by the user ?

Reply:
maybe use same validation from OADP, by importing it

Reply:
I don't think so. If the user will enter invalid key, then BSL created by the user will propagate the error returning in the NaBSL status error conditions. Duplication of this validation doesn't give us a lot of benefit. Note that also if the credentials are valid, but access to S3 is denied there will be as well error coming from the BSL (Velero) which will be propagated back to us same way through BSL.

Reply:
Added as well to the issue #153 I will double check when I will work on that issue, should be minor, however I prefer to double test it.